synapse/storage/databases/main/event_push_actions.py

```diff
@@ -458,57 +473,151 @@ def _get_thread(thread_id: str) -> NotifCounts:
         # updated `event_push_summary` synchronously when persisting a new read
         # receipt).
         txn.execute(
-            """
-                SELECT stream_ordering, notif_count, COALESCE(unread_count, 0), thread_id
+            f"""
+                SELECT notif_count, COALESCE(unread_count, 0), thread_id
                 FROM event_push_summary
+                LEFT JOIN (
+                    SELECT thread_id, MAX(stream_ordering) AS receipt_stream_ordering
+                    FROM receipts_linearized
+                    LEFT JOIN events USING (room_id, event_id)
+                    WHERE
+                        user_id = ?
+                        AND room_id = ?
+                        AND {receipt_types_clause}
+                    GROUP BY thread_id
```

Comment:
There's a handful of queries here which essentially have this same inner query to get the latest receipt per thread for the current room / user.

Is there a better way to do this once instead of continually re-calculating it? Pulling it into Python and then pushing back to the database will cause complicated queries (and could be quite a bit of data to (de)serialize). Could we make a temporary table or something?

Reply:
@erikjohnston suggested there isn't really a better way (besides doing the query first and shipping it all back into Python). But hopes that the disk cache will keep it hot since the queries are done in a row.

See below for more conversation on this: https://github.com/matrix-org/synapse/pull/13877#discussion_r983302762